pkg/controller/periodic/periodic.go

```diff
@@ -98,15 +114,52 @@ func (c *Controller) Gather() {
 		Steps:    threshold,
 		Cap:      interval,
 	}
-	for name := range c.gatherers {
+
+	// flush when all necessary gatherers were processed
+	defer func() {
+		if err := c.recorder.Flush(); err != nil {
+			klog.Errorf("Unable to flush the recorder: %v", err)
+		}
+	}()
+
+	var gatherersToProcess []gather.Interface
+
+	for _, gatherer := range c.gatherers {
+		if lowPriorityGatherer, ok := gatherer.(gather.LowPriorityGatherer); ok {
+			gathererName := lowPriorityGatherer.GetName()
+			lastProcessedTime := c.gatherersLastProcessingTimes[gathererName]
+			timeToProcess := lastProcessedTime.Add(lowPriorityGatherer.GetDesiredGatheringPeriod())
+
+			// only include if it's been at least period since the last processing
+			if time.Now().Equal(timeToProcess) || time.Now().After(timeToProcess) {
+				gatherersToProcess = append(gatherersToProcess, lowPriorityGatherer)
+				c.gatherersLastProcessingTimes[gathererName] = time.Now()
+			}
+		} else {
+			gatherersToProcess = append(gatherersToProcess, gatherer)
+		}
+	}
+
+	var allFunctionReports []gather.GathererFunctionReport
+
+	for _, gatherer := range gatherersToProcess {
 		_ = wait.ExponentialBackoff(backoff, func() (bool, error) {
+			name := gatherer.GetName()
 			start := time.Now()
-			err := c.runGatherer(name)
+
+			ctx, cancel := context.WithTimeout(context.Background(), c.configurator.Config().Interval/2)
+			defer cancel()
+
+			klog.V(4).Infof("Running %s gatherer", gatherer.GetName())
+			var err error
```

Comment:
It looks like we don't need this line :)

Reply:
No idea why it's here :D